tests/e2e/stateapp/stateapp_test.go

```diff
@@ -247,3 +315,54 @@ func TestStateApp(t *testing.T) {
 		})
 	}
 }
+
+func TestStateTransactionApps(t *testing.T) {
+	externalURL := tr.Platform.AcquireAppExternalURL(appName)
+	require.NotEmpty(t, externalURL, "external URL must not be empty!")
+
+	testStateTransactionCase := generateStateTransactionCases()
+	// This initial probe makes the test wait a little bit longer when needed,
+	// making this test less flaky due to delays in the deployment.
+	_, err := utils.HTTPGetNTimes(externalURL, numHealthChecks)
+	require.NoError(t, err)
+
+	// Now we are ready to run the actual tests
+	t.Run("Test HTTP State Transactions", func(t *testing.T) {
```

Comment:
two sub-tests are almost same code base. Do you have an idea to remove the duplicated codes?

for example, you can leverage [table driven tests](https://github.com/golang/go/wiki/TableDrivenTests#example-of-a-table-driven-test) here.

```
var transactionTests = []struct{
  protocol string
  in testStateTransactionCase
}{
  {"HTTP", testStateTransactionCase},
  {"GRPC", testStateTransactionCase},
}

for _, tt := range flagtests {
  t.Run(fmt.Sprintf("state transaction - %s", tt.protocol), func(t *testing.T) {
      for _, step := range tt.in.steps {
                         body, err := json.Marshal(step.request)
			require.NoError(t, err)

			var url string
			if step.command == "get" || tt.protocol == "HTTP" {
				url = fmt.Sprintf("%s/test/%s", externalURL, step.command)
			} else {
        ....
      }
   }
}
```